File: tests/testselections.py
#!/usr/bin/env python
from basetest import BaseTest
from io import BytesIO
import sys, os
import unittest

sys.path.insert(0, '..')
from zeroinstall.injector import selections, namespaces, qdom

mydir = os.path.dirname(os.path.abspath(__file__))
runexec = os.path.join(mydir, 'runnable', 'RunExec.xml')
runnable = os.path.join(mydir, 'runnable', 'Runnable.xml')

class TestSelections(BaseTest):
	def testSelections(self):
		source = self.config.iface_cache.get_interface('http://foo/Source.xml')
		compiler = self.config.iface_cache.get_interface('http://foo/Compiler.xml')
		self.import_feed(source.uri, 'Source.xml')
		self.import_feed(compiler.uri, 'Compiler.xml')

		out, err = self.run_ocaml(['select', '--xml', '--offline', '--command=compile', '--source', 'http://foo/Source.xml'], binary = True)
		assert not err, err
		sels = selections.Selections(qdom.parse(BytesIO(out)))

		def assertSel(s):
			self.assertEqual('http://foo/Source.xml', s.interface)
			self.assertEqual(2, len(s.selections))

			sels = [(sel.interface, sel) for sel in s.selections.values()]
			sels.sort()
			sels = [sel for uri,sel in sels]
			
			self.assertEqual('http://foo/Compiler.xml', sels[0].interface)
			self.assertEqual('http://foo/Source.xml', sels[1].interface)

			self.assertEqual("sha1=345", sels[0].id)
			self.assertEqual("1.0", sels[0].version)

			self.assertEqual('sha1=234', sels[1].id)
			self.assertEqual("1.0", sels[1].version)
			self.assertEqual("bar", sels[1].attrs['http://namespace foo'])
			self.assertEqual("1.0", sels[1].attrs['version'])
			assert 'version-modifier' not in sels[1].attrs

			self.assertEqual(0, len(sels[0].bindings))
			self.assertEqual(0, len(sels[0].dependencies))

			self.assertEqual(3, len(sels[1].bindings))
			self.assertEqual('.', sels[1].bindings[0].insert)
			self.assertEqual('/', sels[1].bindings[1].mount_point)
			self.assertEqual('source', sels[1].bindings[2].qdom.attrs['foo'])

			self.assertEqual(1, len(sels[1].dependencies))
			dep = sels[1].dependencies[0]
			self.assertEqual('http://foo/Compiler.xml', dep.interface)
			self.assertEqual(4, len(dep.bindings))
			self.assertEqual('bin', dep.bindings[0].insert)
			self.assertEqual('PATH', dep.bindings[0].name)
			self.assertEqual('prepend', dep.bindings[0].mode)
			assert dep.bindings[0].separator in ';:'

			self.assertEqual('bin', dep.bindings[1].value)
			self.assertEqual('NO_PATH', dep.bindings[1].name)
			self.assertEqual(',', dep.bindings[1].separator)

			self.assertEqual('bin', dep.bindings[2].insert)
			self.assertEqual('BINDIR', dep.bindings[2].name)
			self.assertEqual('replace', dep.bindings[2].mode)

			foo_binding = dep.bindings[3]
			self.assertEqual('compiler', foo_binding.qdom.attrs['foo'])
			self.assertEqual('child', foo_binding.qdom.childNodes[0].name)
			self.assertEqual('run', foo_binding.command)

			self.assertEqual(["sha1=345", 'sha256new_345'], sorted(sels[0].digests))

		s1 = sels
		s1.selections['http://foo/Source.xml'].attrs['http://namespace foo'] = 'bar'
		assertSel(s1)

		xml = s1.toDOM().toxml("utf-8")
		root = qdom.parse(BytesIO(xml))
		self.assertEqual(namespaces.XMLNS_IFACE, root.uri)

		s2 = selections.Selections(root)
		assertSel(s2)
	
	def testLocalPath(self):
		iface = os.path.join(mydir, "Local.xml")
		out, err = self.run_ocaml(['select', '--xml', iface], binary = True)
		assert not err, err
		s1 = selections.Selections(qdom.parse(BytesIO(out)))
		xml = s1.toDOM().toxml("utf-8")

		# Reload selections and check they're the same
		root = qdom.parse(BytesIO(xml))
		s2 = selections.Selections(root)
		local_path = s2.selections[iface].local_path
		assert os.path.isdir(local_path), local_path
		assert not s2.selections[iface].digests, s2.selections[iface].digests

		iface = os.path.join(mydir, "Local2.xml")
		# Add a newer implementation and try again
		out, err = self.run_ocaml(['select', '--xml', iface], binary = True)
		assert not err, err
		s1 = selections.Selections(qdom.parse(BytesIO(out)))
		#assert s1.get_unavailable_selections(self.config, True)

		xml = s1.toDOM().toxml("utf-8")
		root = qdom.parse(BytesIO(xml))
		s2 = selections.Selections(root)
		xml = s2.toDOM().toxml("utf-8")
		qdom.parse(BytesIO(xml))
		assert s2.selections[iface].local_path is None
		assert not s2.selections[iface].digests, s2.selections[iface].digests
		assert s2.selections[iface].id == 'foo bar=123'

	def testCommands(self):
		iface = os.path.join(mydir, "Command.xml")
		out, err = self.run_ocaml(['select', '--xml', iface], binary = True)
		assert not err, err
		s1 = selections.Selections(qdom.parse(BytesIO(out)))

		impl = s1.selections[iface]
		assert impl.id == 'c'
		assert impl.get_command('run').path == 'test-gui', impl

		dep_impl_uri = impl.get_command('run').requires[0].interface
		dep_impl = s1.selections[dep_impl_uri]
		assert dep_impl.id == 'sha1=256'

		assert s1.commands[0].path == 'test-gui'
		xml = s1.toDOM().toxml("utf-8")
		root = qdom.parse(BytesIO(xml))
		s2 = selections.Selections(root)

		assert s2.commands[0].path == 'test-gui'
		impl = s2.selections[iface]
		assert impl.id == 'c'

		assert s2.commands[0].qdom.attrs['http://custom attr'] == 'namespaced'
		assert len([node for node in s2.commands[0].qdom.childNodes if node.name == 'child']) == 1

		dep_impl = s2.selections[dep_impl_uri]
		assert dep_impl.id == 'sha1=256'

		out, err = self.run_ocaml(['download', '--offline', '--xml', runexec], binary = True)
		assert not err, err
		sels = selections.Selections(qdom.parse(BytesIO(out)))

		xml = sels.toDOM().toxml("utf-8")
		root = qdom.parse(BytesIO(xml))
		s3 = selections.Selections(root)
		runnable_impl = s3.selections[runnable]
		assert 'foo' in runnable_impl.commands
		assert 'run' in runnable_impl.commands

	def testOldCommands(self):
		command_feed = os.path.join(mydir, 'old-selections.xml')
		with open(command_feed, 'rb') as stream:
			s1 = selections.Selections(qdom.parse(stream))
		self.assertEqual("run", s1.command)
		self.assertEqual(2, len(s1.commands))
		self.assertEqual("bin/java", s1.commands[1].path)

		xml = s1.toDOM().toxml("utf-8")
		root = qdom.parse(BytesIO(xml))
		s2 = selections.Selections(root)

		self.assertEqual("run", s2.command)
		self.assertEqual(2, len(s2.commands))
		self.assertEqual("bin/java", s2.commands[1].path)


if __name__ == '__main__':
	unittest.main()

File: ocaml/tests/test_selections.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

open Support.Common
open OUnit
open Zeroinstall
open Zeroinstall.General

module Q = Support.Qdom
module U = Support.Utils

let assert_str_equal = Fake_system.assert_str_equal

let get_sels config fake_system uri =
  fake_system#set_argv [| Test_0install.test_0install; "select"; "--xml"; uri|];
  let output = Fake_system.capture_stdout (fun () -> Main.main config.system) in
  `String (0, output) |> Xmlm.make_input |> Q.parse_input None |> Selections.create

let suite = "selections">::: [
  "selections">:: Fake_system.with_fake_config (fun (config, fake_system) ->
    let import name =
      let url = `remote_feed ("http://foo/" ^ name) in
      U.copy_file config.system (Test_0install.feed_dir +/ name) (Test_driver.cache_path_for config url) 0o644;
      Zeroinstall.Feed.update_last_checked_time config url in
    import "Source.xml";
    import "Compiler.xml";
    fake_system#set_argv [| Test_0install.test_0install; "select"; "--xml";  "--command=compile"; "--source"; "http://foo/Source.xml" |];
    let output = Fake_system.capture_stdout (fun () -> Main.main config.system) in
    let old_sels = `String (0, output) |> Xmlm.make_input |> Q.parse_input None |> Selections.create in
    let index = Selections.make_selection_map old_sels in
    let source_sel = StringMap.find_safe "http://foo/Source.xml" index in
    Q.set_attribute_ns ~prefix:"foo" ("http://namespace", "foo") "bar" source_sel;

    (* Convert to string and back to XML to check we don't lose anything. *)
    let old_xml = Selections.as_xml old_sels in
    let as_str = Q.to_utf8 old_xml in
    let s = `String (0, as_str) |> Xmlm.make_input |> Q.parse_input None in

    ZI.check_tag "selections" s;
    assert_equal 0 @@ Q.compare_nodes ~ignore_whitespace:false old_xml s;

    assert_equal "http://foo/Source.xml" @@ ZI.get_attribute "interface" s;

    let index = Selections.make_selection_map (Selections.create s) in
    assert_equal 2 @@ StringMap.cardinal index;

    let ifaces = StringMap.bindings index |> List.map fst in
    Fake_system.equal_str_lists ["http://foo/Compiler.xml"; "http://foo/Source.xml"] @@ ifaces;

    match StringMap.bindings index |> List.map snd with
    | [comp; src] -> (
        assert_str_equal "sha1=345" @@ ZI.get_attribute "id" comp;
        assert_str_equal "1.0" @@ ZI.get_attribute "version" comp;

        assert_str_equal "sha1=3ce644dc725f1d21cfcf02562c76f375944b266a" @@ ZI.get_attribute "id" src;
        assert_str_equal "1.0" @@ ZI.get_attribute "version" src;
        src.Q.attrs |> Q.AttrMap.get ("http://namespace", "foo") |> Fake_system.expect |> assert_str_equal "bar";
        assert_str_equal "1.0" @@ ZI.get_attribute "version" src;
        src.Q.attrs |> Q.AttrMap.get_no_ns "version-modifier" |> assert_equal None;

        let comp_bindings = comp |> ZI.filter_map Binding.parse_binding in
        let comp_deps = Selections.get_dependencies ~restricts:true comp in
        assert_equal [] @@ comp_bindings;
        assert_equal [] @@ comp_deps;

        let open Binding in

        let () =
          match src |> ZI.filter_map Binding.parse_binding with
          | [EnvironmentBinding {mode = Replace; source = InsertPath "."; _};
             GenericBinding b;
             GenericBinding c] ->
               assert_str_equal "/" @@ ZI.get_attribute "mount-point" b;
               assert_str_equal "source" @@ ZI.get_attribute "foo" c;
          | _ -> assert false in

        let () =
          match Selections.get_source comp with
          | Selections.CacheSelection [("sha256new", "RPUJPVVHEWJ673N736OCN7EMESYAEYM2UAY6OJ4MDFGUZ7QACLKA"); ("sha1", "345")] -> ()
          | _ -> assert false in

        match Selections.get_dependencies ~restricts:true src with
        | [dep] -> (
            assert_str_equal "http://foo/Compiler.xml" @@ ZI.get_attribute "interface" dep;
            match dep |> ZI.filter_map Binding.parse_binding with
            | [EnvironmentBinding {var_name = "PATH"; mode = Add {separator; _}; source = InsertPath "bin"};
               EnvironmentBinding {var_name = "NO_PATH"; mode = Add {separator = ","; _}; source = Value "bin"};
               EnvironmentBinding {var_name = "BINDIR"; mode = Replace; source = InsertPath "bin"};
               GenericBinding foo_binding] -> (
                 assert (separator = ";" || separator = ":");

                 assert_str_equal "compiler" @@ ZI.get_attribute "foo" foo_binding;
                 assert_equal (Some "child") @@ ZI.tag @@ List.hd foo_binding.Q.child_nodes;
                 assert_str_equal "run" @@ ZI.get_attribute "command" foo_binding;
               )
          | _ -> assert false
        )
        | _ -> assert false
    )
    | _ -> assert false
  );

  "local-path">:: Fake_system.with_fake_config (fun (config, fake_system) ->
    let iface = Test_0install.feed_dir +/ "Local.xml" in

    let index = Selections.make_selection_map (get_sels config fake_system iface) in
    let sel = StringMap.find_safe iface index in
    let () =
      match Selections.get_source sel with
      | Selections.LocalSelection local_path ->
          assert (fake_system#file_exists local_path);
      | _ -> assert false in

    let iface = Test_0install.feed_dir +/ "Local2.xml" in
    (* Add a newer implementation and try again *)
    let sels = get_sels config fake_system iface in
    let index = Selections.make_selection_map sels in
    let sel = StringMap.find_safe iface index in
    let tools = Fake_system.make_tools config in
    assert (Driver.get_unavailable_selections ~distro:tools#distro config sels <> []);

    let () =
      match Selections.get_source sel with
      | Selections.CacheSelection [("sha1", "999")] -> ()
      | _ -> assert false in

    assert_equal Feed_url.({id = "foo bar=123"; feed = `local_feed iface}) @@ Selections.get_id sel
  );

  "commands">:: Fake_system.with_fake_config (fun (config, fake_system) ->
    let iface = Test_0install.feed_dir +/ "Command.xml" in
    let sels = get_sels config fake_system iface in
    let index = Selections.make_selection_map sels in
    let sel = StringMap.find_safe iface index in

    assert_equal "c" @@ ZI.get_attribute "id" sel;
    let run = Command.get_command_ex "run" sel in
    assert_equal "test-gui" @@ ZI.get_attribute "path" run;
    run.Q.attrs |> Q.AttrMap.get ("http://custom", "attr") |> assert_equal (Some "namespaced");
    assert_equal 1 @@ List.length (run.Q.child_nodes |> List.filter (fun node -> snd node.Q.tag = "child"));

    let () =
      match Selections.get_dependencies ~restricts:true run with
      | dep :: _ ->
          let dep_impl_uri = ZI.get_attribute "interface" dep in
          let dep_impl = StringMap.find_safe dep_impl_uri index in
          assert_equal "sha1=256" @@ ZI.get_attribute "id" dep_impl;
      | _ -> assert false in

    let runexec = Test_0install.feed_dir +/ "runnable" +/ "RunExec.xml" in
    let runnable = Test_0install.feed_dir +/ "runnable" +/ "Runnable.xml" in

    fake_system#set_argv [| Test_0install.test_0install; "download"; "--offline"; "--xml"; runexec|];
    let output = Fake_system.capture_stdout (fun () -> Main.main config.system) in
    let s3 = `String (0, output) |> Xmlm.make_input |> Q.parse_input None |> Selections.create in
    let index = Selections.make_selection_map s3 in

    let runnable_impl = StringMap.find_safe runnable index in
    runnable_impl |> ZI.filter_map (fun child ->
      if ZI.tag child = Some "command" then Some (ZI.get_attribute "name" child) else None
    ) |> Fake_system.equal_str_lists ["foo"; "run"]
  );

  "old-commands">:: Fake_system.with_fake_config (fun (config, _fake_system) ->
    let command_feed = Test_0install.feed_dir +/ "old-selections.xml" in
    let sels = Q.parse_file config.system command_feed |> Selections.create in

    let user_store = List.hd config.stores in
    let add_impl digest_str =
      U.makedirs config.system (user_store +/ digest_str) 0o755 in

    add_impl "sha1new=daf7bfada93ec758baeef1c714f3239ce0a5a462";
    add_impl "sha1new=3ede8dc4b83dd3d7705ee3a427b637a2cb98d789";
    add_impl "sha256=d5f30349df0fac73c4621cc3161ab125ba5f83ba9ec35e27fbb0f3a7392070eb";

    let (args, _env) = Exec.get_exec_args {config with dry_run = true} sels [] in
    match args with
    | [_; "eu.serscis.Eval"] -> ()
    | _ -> assert false
  );
]
